CASSANDRA-15792 Fix read_repair_test.TestSpeculativeReadRepair test_speculative_data_request

grighetto opened the pull request:
CLOSED IN FAVOR OF #2 

-----------------------
Use a longer write_request_timeout_in_ms in read repair test that expects no speculated writes.

adelapena committed:
Add tests for CASSANDRA-8272 and CASSANDRA-8273

jolynch committed:
Update dtests for optional server TLS changes

Patch by Joseph Lynch; Reviewed by Ekaterina Dimitrova for CASSANDRA-15262

snazy committed:
CASSANDRA-15835: use correct Java version for upgrade-tests

snazy committed:
CASSANDRA-15835: re-enable running upgrade-dtests for developer-CI runs against a local C* clone

snazy committed:
CASSANDRA-15835: don't let check_logs_for_errors() fail, if there's no log file

snazy committed:
CASSANDRA-15835: also copy the new ccm node startup-stdout/stderr.log files

grighetto committed:
CASSANDRA-15792 Use a longer write_request_timeout_in_ms in read repair test that expects no speculated writes.

grighetto committed:
CASSANDRA-15792 Use a longer write_request_timeout_in_ms in read repair test that expects no speculated writes.

grighetto committed:
CASSANDRA-15792 Made TestSpeculativeReadRepair.test_speculative_data_request deterministic by relaxing the assertion around speculated writes. If a speculated write is detected, then we confirm the repair messages were sent in the expected order (first to node 3 then to node 2).

grighetto committed:
CASSANDRA-15792 Use a longer write_request_timeout_in_ms in read repair test that expects no speculated writes.

grighetto commented on dtest_setup.py:
```diff
@@ -253,6 +253,9 @@ def patient_cql_connection(self, node, keyspace=None,
         return session
 
+    def get_cql_connection(self, node, **kwargs):
+        return self.patient_exclusive_cql_connection(node, retry_policy=None, **kwargs)
```
this utility function was defined twice in two subclasses, so moved it up to the parent to make it reusable

grighetto commented on read_repair_test.py:
```diff
@@ -339,18 +339,18 @@ def test_gcable_tombstone_resurrection_on_range_slice_query(self):
-            for t in trace.events:
-                print(("%s\t%s\t%s\t%s" % (t.source, t.source_elapsed, t.description, t.thread_name)))
-            print(("-" * 40))
+def pprint_trace(trace):
```
moved it out of a class scope to make it reusable throughout the test file

grighetto commented on read_repair_test.py:
```diff
@@ -530,7 +527,10 @@ def test_speculative_data_request(self):
+            # depending on how quickly the read-repair mutation is performed by node 3,
+            # the coordinator may send a speculative write to node 2,
+            # so we accept 0 or 1 speculated write here
+            assert storage_proxy.speculated_rr_write >= 0
```
this fixes the non-deterministic behavior

bereng replied:
Hi, quick suggestion.

Would it make sense to have a counter on the timeout so you could check `assert storage_proxy.speculated_rr_write == timeoutCounter` or similar? That way it would be deterministic both with timeout and without. That would also allow you to remove the added extra test which on an enough loaded system, as you rightfully said, might give trouble. Hope it makes sense :-)

grighetto replied:
@bereng and I discussed this offline, but while it makes to check the timeout happened before the speculated write, the way the code is structured right now the proposed timeout counter would have to be incremented right before the speculated write metric (meaning, right in the line above it), so they would actually be redundant.

But thinking through it again, we hashed out another idea, which is checking if the coordinator node got a response back from node 3, in which case it shouldn't have sent a speculated write to node 2.
On the other hand, if it did send a speculated write, then we have to assert it didn't get a response back in time from node 3.

This way the test becomes deterministic and we can handle both scenarios.

grighetto commented on read_repair_test.py:
```diff
@@ -794,6 +791,61 @@ def test_atomic_writes(self, repair_type, expect_atomic):
+        cluster.set_configuration_options(values={'hinted_handoff_enabled': False,
+                                                  'dynamic_snitch': False,
+                                                  # longer timeout to reduce the likelihood of a speculative write
+                                                  'write_request_timeout_in_ms': 4000,
```
longer write timeout to avoid speculative writes

grighetto commented on read_repair_test.py:
```diff
@@ -794,6 +791,61 @@ def test_atomic_writes(self, repair_type, expect_atomic):
+        session.execute("CREATE TABLE ks.tbl (k int, c int, v int, primary key (k, c)) WITH speculative_retry = '400ms';")
+
+    @since('4.0')
+    def test_speculative_data_request_without_speculative_write(self):
```
replicated the above test here, but in this class the cluster is set up with a longer write_request_timeout_in_ms, because we want to assert that speculated_rr_write is equal to 0

Commit:
Introduce --execute-upgrade-tests-only to only run upgrade tests

snazy committed:
Fix `@since` for upgrade_internal_auth_test.TestAuthUpgrade.test_upgrade_to_22 + test_upgrade_to_30

Patch by Robert Stupp; reviewed by Eduard Tudenhöfner for CASSANDRA-15853

snazy committed:
Merge pull request #71 from snazy/15835-fix-upgrade-dtests

15835 fix upgrade dtests

bereng committed:
CASSANDRA-15858 Fix repair failing tests

bereng committed:
CASSANDRA-15855 Mark flaky until root cause is solved

bereng committed:
Fix flaky replace address tests and bootstrap resume fixes

Patch by Berenguer Blasi, reviewed by brandonwilliams for CASSANDRA-15863

bereng commented:
@grighetto are you waiting or do you need anything from me here?

grighetto commented:
@bereng not waiting, I moved the ticket back to In Progress and I'm working on the implementation right now. I'll ping you for another review when ready.

blerer committed:
Fix describe tests

BrynCooke committed:
Updated tests to expect topology events for localhost clusters Post 4.0.

Patch by Bryn Cooke, reviewed by brandonwilliams for CASSANDRA-15677

bereng committed:
CASSANDRA-15885 Fix flaky dtest assert sorting order

blerer committed:
Merge pull request #80 from blerer/CASSANDRA-14825

Fix describe tests

yifan-c committed:
Add test for CASSANDRA-15905

Patch by Yifan Cai, reviewed by brandonwilliams for CASSANDRA-15905

grighetto committed:
CASSANDRA-15792 Made TestSpeculativeReadRepair.test_speculative_data_request deterministic by relaxing the assertion around speculated writes. If a speculated write is detected, then we confirm the repair messages were sent in the expected order (first to node 3 then to node 2).

grighetto committed:
CASSANDRA-15792 Made TestSpeculativeReadRepair.test_speculative_data_request deterministic by relaxing the assertion around speculated writes. If a speculated write is detected, then we confirm the repair messages were sent in the expected order (first to node 3 then to node 2).

grighetto committed:
Merge branch '15792-fix-test-speculative-data-request' of https://github.com/grighetto/cassandra-dtest into 15792-fix-test-speculative-data-request

# Conflicts:
#	read_repair_test.py

bereng commented:
@grighetto maybe you want to rebase/squash this to sort out all that extra commit noise?

grighetto committed:
Merge branch '15792-fix-test-speculative-data-request' of https://github.com/grighetto/cassandra-dtest into 15792-fix-test-speculative-data-request

grighetto committed:
Merge branch '15792-fix-test-speculative-data-request' of https://github.com/grighetto/cassandra-dtest into 15792-fix-test-speculative-data-request

grighetto commented:
@bereng I rebased and squashed the commits but for some reason the github UI still shows a larger diff to me, sometimes the UI lags behind the actual state, not sure what it looks like for you.
Here's a clean diff though: https://github.com/apache/cassandra-dtest/compare/master...grighetto:15792-fix-test-speculative-data-request
